mlflow/metrics/genai/genai_metric.py

```diff
@@ -558,4 +604,104 @@ def eval_fn(
+            error_code=PUBLIC_UNDOCUMENTED,
+        )
+
+    if mlflow_version_at_ser != VERSION:
+        warnings.warn(
+            f"The custom metric definitions were serialized using MLflow {mlflow_version_at_ser}. "
+            f"Deserializing them with the current version {VERSION} might cause mismatches. "
+            "Please ensure compatibility or consider regenerating the metrics "
+            "using the current version.",
+            UserWarning,
+            stacklevel=2,
+        )
+
+    args_dict.pop("mlflow_version")
+    args_dict.pop("fn_name")
+    if fn_name == make_genai_metric_from_prompt.__name__:
+        return make_genai_metric_from_prompt(**args_dict)
+
+    examples = args_dict["examples"]
+    if examples is None:
+        return make_genai_metric(**args_dict)
+
+    deser_examples = []
+    for example in examples:
+        deser_examples.append(EvaluationExample(**example))
+
+    args_dict["examples"] = deser_examples
+    return make_genai_metric(**args_dict)
+
+
+def search_custom_metrics(
+    run_id: str,
+    name: Optional[str] = None,
+    version: Optional[str] = None,
+) -> List[EvaluationMetric]:
+    """
+    Searches for custom metrics created by users through `make_genai_metric()` or
+    `make_genai_metric_from_prompt()` that are associated with a particular evaluation run.
+
+    Args:
+        run_id: The unique identifier for the run.
+        name: (Optional) The name of the custom metric to search for.
+            If None, search for all metrics.
+        version: (Optional) The version of the custom metric to search for.
+            If None, search for all versions.
+
+    Returns:
+        A list of EvaluationMetric objects that match the search criteria.
+
+    .. code-block:: python
+        :test:
+        :caption: Example for searching for a custom genai metric
+
+        from mlflow.metrics.genai import search_custom_metrics
+
+        metrics = search_custom_metrics(
+            run_id=run.info.run_id, name="answer_similarity", version="v1"
+        )
+    """
+    client = mlflow.MlflowClient()
+    artifacts = [a.path for a in client.list_artifacts(run_id)]
+    if _GENAI_CUSTOM_METRICS_FILE_NAME not in artifacts:
+        _logger.warning("No custom metric definitions were found for this evaluation run.")
+        return []
+
+    with TemporaryDirectory() as tmpdir:
+        downloaded_artifact_path = mlflow.artifacts.download_artifacts(
+            run_id=run_id,
+            artifact_path=_GENAI_CUSTOM_METRICS_FILE_NAME,
+            dst_path=tmpdir,
+        )
+        custom_metrics = client._read_from_file(downloaded_artifact_path)
+
+    if name is not None:
+        custom_metrics = _filter_by_field(custom_metrics, "name", name)
+    if version is not None:
+        custom_metrics = _filter_by_field(custom_metrics, "version", version)
+    metric_args_list = custom_metrics["metric_args"].tolist()
+    if len(metric_args_list) == 0:
```

Comment:
nit: Shall we emit similar warning as L685 here?

Reply:
oh conceptually, trying to search for custom metrics for a run that doesn't contain it at all vs no matched results after name/version filter are different. but yeah it doesn't hurt to emit the warning. I see the point too. Let me add it.